api/src/main/java/org/pmiops/workbench/db/dao/WorkspaceServiceImpl.java

```diff
@@ -0,0 +1,106 @@
+package org.pmiops.workbench.db.dao;
+
+import java.util.List;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.logging.Logger;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import org.pmiops.workbench.db.model.Workspace;
+import org.pmiops.workbench.db.model.WorkspaceUserRole;
+import org.pmiops.workbench.exceptions.BadRequestException;
+import org.pmiops.workbench.exceptions.NotFoundException;
+
+/**
+ * Workspace manipulation and shared business logic which can't be represented by automatic query
+ * generation in WorkspaceDao, or convenience aliases.
+ *
+ * TODO(RW-215) Add versioning to detect/prevent concurrent edits.
+ */
+@Service
+public class WorkspaceServiceImpl implements WorkspaceService {
+  private static final Logger log = Logger.getLogger(WorkspaceService.class.getName());
+
+  /**
+   * Clients wishing to use the auto-generated methods from the DAO interface may directly access
+   * it here.
+   */
+  @Autowired private WorkspaceDao workspaceDao;
+
+  public WorkspaceServiceImpl() {
+
+  }
+
+  public WorkspaceDao getDao() {
+    return workspaceDao;
+  }
+
+  public Workspace get(String ns, String id) {
+    return workspaceDao.findByWorkspaceNamespaceAndFirecloudName(ns, id);
+  }
+
+  public Workspace getRequired(String ns, String id) {
+    Workspace workspace = get(ns, id);
+    if (workspace == null) {
+      throw new NotFoundException(String.format("Workspace %s/%s not found.", ns, id));
+    }
+    return workspace;
+  }
+
+  public List<Workspace> findForReview() {
+    return workspaceDao.findByApprovedIsNullAndReviewRequestedTrueOrderByTimeRequested();
+  }
+
+  // FIXME @Version instead? Bean instantiation v. @Transactional?
+  //@Transactional
+  public void setResearchPurposeApproved(String ns, String id, boolean approved) {
+    Workspace workspace = getRequired(ns, id);
+    if (workspace.getReviewRequested() == null || !workspace.getReviewRequested()) {
+      throw new BadRequestException(String.format(
+          "No review requested for workspace %s/%s.", ns, id));
+    }
+    if (workspace.getApproved() != null) {
+      throw new BadRequestException(String.format(
+          "Workspace %s/%s already %s.",
+          ns, id, workspace.getApproved() ? "approved" : "rejected"));
+    }
+    workspace.setApproved(approved);
+    workspaceDao.save(workspace);
+  }
+
+  @Transactional
+  public void updateUserRoles(String ns, String id, Set<WorkspaceUserRole> userRoleSet) {
+    org.pmiops.workbench.db.model.Workspace dbWorkspace = getRequired(
+        ns, id);
+    Iterator<WorkspaceUserRole> dbUserRoles = dbWorkspace.getWorkspaceUserRoles().iterator();
+    while (dbUserRoles.hasNext()) {
```

Comment:
`for(WorkspaceUserRole dbUserRole : dbWorkspace.getWorkspaceUserRoles())` ? Here and elsewhere, preferred to getting iterators.

Reply:
Got it.

Reply:
Regarding this, @markfickett do you mind explaining why it is preferred to getting iterators? Especially because I learned today that iterators have a nice clean remove function.

Reply:
If you were going to use iterator.remove(), that might be a case for it. Otherwise, for (Foo foo : fooList) is a lot shorter and easier to read.

Reply:
I actually am now using iterator.remove(). (I wasn't using it before and it was causing some messy errors, but I moved to using it and it works cleanly now.)